### DISCUSSION TITLE
(Probably) the ultimate solution to installation failures on Mac OS

### DISCUSSION THREAD

**Discussion by kwt-klure** [networkit/networkit#686] [Show and tell]
I have faced more than a few _seemingly random_ installation failures on Mac OS X for years. I finally find out the reason why there would be failures.

The failure is because of that Mac OS uses Apple LLVM (current) version 10.0.1 (clang-1001.0.46.4) as default compiler. Installing Xcode command line tools and GCC alone cannot solve the problem fundamentally, since the system would still use clang to compile while installing networkit.

Theoretically, we could solve it by changing or cancelling the system default compiler. However, it is not a handy task and sometimes risky. 

Thanks to @sol-prog, who wrote a [guide to compiling GCC9 on macOS Mojave](https://solarianprogrammer.com/2017/05/21/compiling-gcc-macos/). Following the guide, we can solve the problem once and for all (hopefully). Note that you might have to `sudo make` and `sudo configure` depending on the security settings.

You may try this if you already tried the followings but still does not work.

As on 'Get Started' state:

_Mac OS X 10.10 “Yosemite”: Some users have reported compilation problems on Yosemite with g++ 4.9. The compiler errors mention register problems. While the exact reason remains unclear, the actual issue seems to be that the compiler tries to perform a dual architecture build. Fix: Enforce a 64-bit build by prepending `ARCHFLAGS="-arch x86_64"` to your setup/pip command, e.g. as in `sudo ARCHFLAGS="-arch x86_64" python3 setup.py build_ext --inplace -j4` or `sudo ARCHFLAGS="-arch x86_64" pip3 install networkit`._
(By the way, I was one of those "some users" then.)

or

the way mentioned in [https://github.com/networkit/networkit/issues/172](https://github.com/networkit/networkit/issues/172) by @PepSalehi, i.e., `conda install -c vgauthier networkit`.

**Comment by hmeyerhenke** [comment]
Thanks a lot!

Installing GCC9 might be easier with a package manager such as homebrew, though. Our Mac installations of NetworKit have worked with it recently. But the compiler has to be specified explicitly (otherwise the system uses clang).
